pkg/kube/informerfactory/factory.go

```diff
@@ -0,0 +1,222 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package informerfactory provides a "factory" to generate informers. This allows users to create the
+// same informers in multiple different locations, while still using the same underlying resources.
+// Additionally, aggregate operations like Start, Shutdown, and Wait are available.
+// Kubernetes core has informer factories with very similar logic. However, this has a few problems that
+// spurred a fork:
+// * Factories are per package. That means we have ~6 distinct factories, which makes management a hassle.
+// * Across these, the factories are often inconsistent in functionality. Changes to these takes >4 months.
+// * Lack of functionality we want (see below).
+//
+// Added functionality:
+// * Single factory for any type, including dynamic informers, meta informers, typed informers, etc.
+// * Ability to create multiple informers of the same type but with different filters.
+// * Ability to run a single informer rather than all of them.
+package informerfactory
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/istio/pkg/kube/kubetypes"
+	"istio.io/istio/pkg/util/sets"
+)
+
+// NewInformerFunc returns a SharedIndexInformer.
+type NewInformerFunc func() cache.SharedIndexInformer
+
+type StartableInformer struct {
+	Informer cache.SharedIndexInformer
+	start    func(stopCh <-chan struct{})
+}
+
+func (s StartableInformer) Start(stopCh <-chan struct{}) {
+	s.start(stopCh)
+}
+
+// InformerFactory provides access to a shared informer factory
+type InformerFactory interface {
+	// Start initializes all requested informers. They are handled in goroutines
+	// which run until the stop channel gets closed.
+	Start(stopCh <-chan struct{})
+
+	// InformerFor returns the SharedIndexInformer the provided type.
+	InformerFor(resource schema.GroupVersionResource, opts kubetypes.InformerOptions, newFunc NewInformerFunc) StartableInformer
+
+	// WaitForCacheSync blocks until all started informers' caches were synced
+	// or the stop channel gets closed.
+	WaitForCacheSync(stopCh <-chan struct{}) bool
+
+	// Shutdown marks a factory as shutting down. At that point no new
+	// informers can be started anymore and Start will return without
+	// doing anything.
+	//
+	// In addition, Shutdown blocks until all goroutines have terminated. For that
+	// to happen, the close channel(s) that they were started with must be closed,
+	// either before Shutdown gets called or while it is waiting.
+	//
+	// Shutdown may be called multiple times, even concurrently. All such calls will
+	// block until all goroutines have terminated.
+	Shutdown()
```

Comment:
is it equal to stop channel closed?

Reply:
No, its actually waiting for them to completely exit. The naming is weird (copied from upstream), it should be "WaitForTermination" or something. The stop channel needs to be closed independently.

We use it for a few tests to ensure things are cleaned up to avoid race conditions